### ISSUE TITLE
Unhandled rejection overtakes synchronous exception through done()

### ISSUE THREAD

**Issue by rogierschouten** [petkaantonov/bluebird#471]
Seen with Bluebird 2.9.3. 

The behavior below caused a programming error to become hidden, making debugging difficult. The done() method will terminate the application on a thrown error within a promise chain, but it might wait to do so for quite some time, even processing more promises. In the meantime, the application continues to run, generating all kinds of unexpected side effects - after all, we're already in an error state. 

I'm aware that the whole thing is a-synchronous and that you cannot terminate the application immediately on an exception. However, it seems that arbitrarily long promise chains are still executed before the done() - they seem to overtake each other. I would at least expect some execution order to be preserved.

What happened in reality (below is a smaller example) is that an assertion error occurred within a promise chain that ended in .done(). After that, the application was in a state where it behaved wrongly and eventually caused another un-handled rejection. The un-handled rejection overtook the assertion error and only the rejection was shown, obscuring the real cause. Note that the rejection was quite far removed in terms of nodejs cycles.

The example below shows the problem

```javascript
var Promise = require("bluebird");

// kill the process on unhandled rejections
Promise.onPossiblyUnhandledRejection(function(reason) {
        // do some custom logging
	console.log("possibly unhandled rejected Promise", reason);
	process.exit(1);
});

Promise.resolve()
	.then(function() {
		// longer chain expected to give error long after the process is dead
                // not because .done() but because of possibly unhandled rejection handler.
		Promise.resolve()
			.then(function() {				
				Promise.resolve()
					.then(function() {				
						Promise.reject(new Error("unhandled rejection"));
					});
			});
		
		// exception expected to terminate the process
		throw new Error("unhandled thrown error which should end the process because of done()");
	})
	.done();
```

Console output:
```shell
possibly unhandled rejected Promise [Error: unhandled rejection]
```

Is this expected behavior?

**Comment by spion** [comment]
The problem is in async.js using setTimeout to throw the error, if available - maybe it can be fixed by a patch that rethrows it both synchronously and asynchronously.

**Comment by rogierschouten** [comment]
We also have this: 

```javascript
Promise.setScheduler(function(callback) {
	setImmediate.apply(null, [callback]);
});
```

I would at least have expected that to be used also for .done().   But I agree that throwing synchronously would be even better.

**Comment by petkaantonov** [comment]
I want to throw synchronously but if the application recovers from the exception, it will leave bluebird in inconsistent state

**Comment by rogierschouten** [comment]
Would that matter once you've made a programmer error?

**Comment by petkaantonov** [comment]
In practice, yes.. just open the console of any website, it's full of uncaught errors and the applications appear to be keep working just fine.

**Comment by rogierschouten** [comment]
Could we make it an option, then? So we can have a sync exception server-side, or while not in production?

**Comment by petkaantonov** [comment]
What if in node it did the same as event emitters... write stack trace to console and exit immediately? shouldn't that be effectively the same as throwing synchronously but without the risk that the application recovers?

**Comment by rogierschouten** [comment]
Yes that would be nice. But in that case I would expect a handler similar to onPossiblyUnhandledRejection() so that I can flush the logs.

**Comment by petkaantonov** [comment]
You mean `process.on("exit"`?

**Comment by rogierschouten** [comment]
Good point

**Comment by rogierschouten** [comment]
No wait, that means I cannot log the exception to my own logging system anymore and also I cannot flush the logs because process.on("exit" only allows sync actions.

**Comment by petkaantonov** [comment]
So if you need to do async stuff before the exception can be thrown, it will not be any different from the status quo as while you are waiting for the logs to flush, the unhandled rejection handler will be called and call process.exit

**Comment by petkaantonov** [comment]
@rogierschouten 3.0.0 will be soon released 2-4 weeks, do you have any input? I think you can synchronously flush logs with `writeSync` using stdout file descriptor.

Otherwise this will be left as (as it doesn't change anything if you need to flush logs asynchronously) is and the next time it can be changed is 4.0.

**Comment by rogierschouten** [comment]
Hi Petka,

I agree with you, you can throw synchronously.Thank you.

Rogier

**Comment by petkaantonov** [comment]
Fixed in 3.0 release
